sphinx/util/__init__.py

```diff
@@ -182,6 +182,46 @@ def copy_static_entry(source, targetdir, builder, context={},
                               exclude_matchers=exclude_matchers)
 
 
+def copy_extra_entry(source, targetdir, builder, exclude_matchers=()):
+    """Copy a HTML builder extra_path entry from source to targetdir.
+
+    Handles all possible cases of files, directories and subdirectories.
+    """
+    def excluded(root, entry):
+        relroot = relative_path(os.path.dirname(source), root)
+        path = os.path.join(relroot, entry)
+        for matcher in exclude_matchers:
+            if matcher(path):
+                return True
+        else:
+            return False
+
+    if os.path.isfile(source):
+        if not excluded(source, ''):
+            target = os.path.join(targetdir, os.path.basename(source))
+            copyfile(source, target)
+        return
+
+    excluded_dir = []
+    for root, dirs, files in os.walk(source):
+        if root in excluded_dir:
+            continue
+
+        relroot = relative_path(source, root)
+        for dir in dirs:
+            if excluded(root, dir):
+                excluded_dir.append(os.path.join(root, dir))
+            else:
+                target = os.path.join(targetdir, relroot, dir)
+                if not path.exists(target):
+                    os.mkdir(target)
+        for file in files:
+            if not excluded(root, file):
+                srcpath = os.path.join(root, file)
+                target = os.path.join(targetdir, relroot, file)
+                copyfile(srcpath, target)
+
+
```

Comment:
I'm very sorry. It's my mistake on merging. It works fine.
I never know that changing `dirs` argument from `os.walk()` prunes directory traversals. 

Thank you for proposal. I would merge this soon.